### PR TITLE
Allow starting a console attached to a text editor

### LINKED ISSUE #927: Allow starting a console attached to a text editor
We want to start to allow text files to be associated with kernels and consoles. This issue describes an MVP to get us going in that direction. Here is a description of this MVP:

* Add a text editor command that opens a new console for that editor with a new kernel of the user's choice.
* Once a console has been opened for a file, bind shift+enter to run the selected code from the file in the console.
* When this is done, we should also probably focus/activate that console.
* Eventually, we should add some sort of visual styling to show that the file goes with the console.

### PR DESCRIPTION
**Pull request by blink1073**
Fixes #927.  Fixes #947.

editor-console

### CODE REVIEW

**Comment by blink1073** [comment]
Ready for review!

**Comment by ellisonbg** [comment]
Works for me. Will review code tomorrow.

One behavioral change: if there is no text selected, run the current line upon shift+enter.

File: src/console/plugin.ts [hunk 1]
```diff
@@ -84,10 +88,16 @@ const LANDSCAPE_ICON_CLASS = 'jp-MainAreaLandscapeIcon';
  */
 const CONSOLE_ICON_CLASS = 'jp-ImageConsole';
 
+
 /**
- * The file extension for consoles.
+ * The interface for a start console.
```
Comment (ellisonbg) [inline src/console/plugin.ts]: The phrase "start console" is a verb, but we are using it here as a noun. Maybe rename the class to something that is more noun like? Maybe "ConsoleStarter"?

File: src/console/plugin.ts [hunk 2]
```diff
@@ -255,6 +227,120 @@ function activateConsole(app: JupyterLab, services: IServiceManager, rendermime:
   menu.addItem({ command });
 
 
+  command = 'console:create';
+  commands.addCommand(command, {
+    execute: (args: IStartConsole) => {
```
Comment (ellisonbg) [inline src/console/plugin.ts]: Ahh, or maybe name it `StartConsoleArgs` or `CreateConsoleArgs`

File: src/console/plugin.ts [hunk 3]
```diff
@@ -255,6 +227,120 @@ function activateConsole(app: JupyterLab, services: IServiceManager, rendermime:
...
+      if (args.sessionId) {
+        return manager.connectTo(args.sessionId).then(session => {
+          createConsole(session);
+          return session.id;
+        });
+      }
+
+      // Find the correct path for the new session.
+      // Use the given path or the cwd.
+      let path = args.path || pathTracker.path;
```
Comment (ellisonbg) [inline src/console/plugin.ts]: Very nice that this uses the same path as the text file!

File: src/editorwidget/plugin.ts [hunk 4]
```diff
@@ -80,7 +84,9 @@ const cmdIds = {
   matchBrackets: 'editor:match-brackets',
   vimMode: 'editor:vim-mode',
   closeAll: 'editor:close-all',
-  changeTheme: 'editor:change-theme'
+  changeTheme: 'editor:change-theme',
+  startConsole: 'editor:start-console',
+  runCode: 'editor:run-code'
```
Comment (ellisonbg) [inline src/editorwidget/plugin.ts]: Let's decide is we are going to use "run" or "execute" in the UI and commands and then use the word consistently.

File: src/editorwidget/plugin.ts [hunk 5]
```diff
@@ -80,7 +84,9 @@ const cmdIds = {
   matchBrackets: 'editor:match-brackets',
   vimMode: 'editor:vim-mode',
   closeAll: 'editor:close-all',
-  changeTheme: 'editor:change-theme'
+  changeTheme: 'editor:change-theme',
+  startConsole: 'editor:start-console',
```
Comment (ellisonbg) [inline src/editorwidget/plugin.ts]: Are we using "start" or "create" in the UI/commands. Let's be consistent as it makes it easier for users to remember the names of commands.

File: src/editorwidget/plugin.ts [hunk 6]
```diff
@@ -106,47 +112,92 @@ function activateEditorHandler(app: JupyterLab, registry: IDocumentRegistry, mai
...
     label: 'Toggle Vim Mode'
   });
-  app.commands.addCommand(cmdIds.closeAll, {
+
+  commands.addCommand(cmdIds.closeAll, {
     execute: () => { closeAllFiles(tracker); },
     label: 'Close all files'
   });
+
+  // TODO: add an attached property to the widget with the session id.
```
Comment (ellisonbg) [inline src/editorwidget/plugin.ts]: Do you want to do this before we merge?

**Comment by blink1073** [comment]
All comments addressed.

**Pull request merged**